sdk/storage/inc/blobs/blob_sas_builder.hpp

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <string>
+
+#include "blobs/protocol/blob_rest_client.hpp"
+#include "common/account_sas_builder.hpp"
+
+namespace Azure { namespace Storage { namespace Blobs {
+
+  enum class BlobSasResource
+  {
+    Container,
+    Blob,
+    BlobSnapshot,
+    BlobVersion,
+  };
+
+  enum class BlobContainerSasPermissions
+  {
+    None = 0,
+    Read = 1,
+    Write = 2,
+    Delete = 4,
+    List = 8,
+    Add = 16,
+    Create = 32,
+    Tags = 64,
+    DeleteVersion = 128,
+    All = ~0,
+  };
+
+  inline BlobContainerSasPermissions operator|(
+      BlobContainerSasPermissions lhs,
+      BlobContainerSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<BlobContainerSasPermissions>;
+    return static_cast<BlobContainerSasPermissions>(
+        static_cast<type>(lhs) | static_cast<type>(rhs));
+  }
+
+  inline BlobContainerSasPermissions operator&(
+      BlobContainerSasPermissions lhs,
+      BlobContainerSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<BlobContainerSasPermissions>;
+    return static_cast<BlobContainerSasPermissions>(
+        static_cast<type>(lhs) & static_cast<type>(rhs));
+  }
+
+  enum class BlobSasPermissions
+  {
+    None = 0,
+    Read = 1,
+    Write = 2,
+    Delete = 4,
+    Add = 8,
+    Create = 16,
+    Tags = 32,
+    DeleteVersion = 64,
+    All = ~0,
+  };
+
+  inline BlobSasPermissions operator|(BlobSasPermissions lhs, BlobSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<BlobSasPermissions>;
+    return static_cast<BlobSasPermissions>(static_cast<type>(lhs) | static_cast<type>(rhs));
+  }
+
+  inline BlobSasPermissions operator&(BlobSasPermissions lhs, BlobSasPermissions rhs)
+  {
+    using type = std::underlying_type_t<BlobSasPermissions>;
+    return static_cast<BlobSasPermissions>(static_cast<type>(lhs) & static_cast<type>(rhs));
+  }
+
+  struct BlobSasBuilder
+  {
+    std::string Version = c_APIVersion;
+    SasProtocol Protocol;
+    std::string StartsOn;
+    std::string ExpiresOn;
+    std::string IPRange;
+    std::string Identifier;
+    std::string ContainerName;
+    std::string BlobName;
+    std::string Snapshot;
+    BlobSasResource Resource;
+
+    std::string CacheControl;
+    std::string ContentDisposition;
+    std::string ContentEncoding;
+    std::string ContentLanguage;
+    std::string ContentType;
+
+    void SetPermissions(BlobContainerSasPermissions permissions);
+    void SetPermissions(BlobSasPermissions permissions);
+
+    std::string ToSasQueryParameters(const SharedKeyCredential& credential);
```

Comment:
`SharedKeyCredential`'s constructor takes `std::string accountName, std::string accountKey`. Is it better to make the parameters be `const std::string& sharedKey,
        const std::string& accountName` to be more clear of what is needed and more consistent with the function below?

Reply:
We already have an abstraction of account name and shared key, is there any reason you don't want to use it? Plus key is supposed to be a secret thing. I don't think it's good practice to pass this secret around everywhere.

Reply:
So I got the same question for you. In [filesystem client](https://github.com/Azure/azure-sdk-for-cpp/blob/573ffdf1f73b4a03861dcf1c3fd0a761cce588d0/sdk/storage/inc/datalake/file_system_client.hpp#L48), why don't you have the constructor accept account name and shared key, but use `SharedKeyCredential` instead?

Reply:
The only concern is the overhead introduced by `SharedKeyCredential` because it has mutex in it. It has no additional usage of the class itself besides the fact that it holds the account name and key.

Reply:
As we talked offline, filesystem client holds `SharedKeyCredential` and uses other functionalities provided by the class, including thread safety insurance.

Reply:
i'm not gonna take that as a convincing argument unless you have evidence shows that the performance is impacted.